### PR TITLE
Make syncthing honor the ignorePermissions/FlagNoPermBits settings (fixes #400)

### LINKED ISSUE #400: FreeBSD on FreeNAS: read-only file system
I just tried to install syncthing on a FreeNAS (FreeNAS-9.2.1.5-RELEASE-x64 (80c1d35))

I got the following message, because FreeNAS is usually installed on an USB-stick. 

freenas# ./syncthing
10:56:00 FATAL: mkdir /root/.config: read-only file system

It would be more than nice, if we can call ./syncthing with an option to save the necessary .config file to an other destination.

BTW: 
syncthing --version
Display the version of syncthing.

syncthing -h
syncthing --help
Print a help message describing all of syncthing's command-line options.

...

Nonetheless, thanks for this great application, with I actually run under Windows 7 and 8.1.

Cheers,

merlinuwe

### PR DESCRIPTION
**Pull request by mogwa1**
This is a suspected bug that I ran into while trying to solve issue #1339.

What the code should be doing is:
- According to https://forum.syncthing.net/t/v0-8-10-ignore-permissions/263/3, the ignorePermissions flag should only be picked up on the server where a new file is created. If ignorePermissions is set on that server, the FlagNoPermBits is added to the file flags.
- On the receiving side, the file flags should be checked for FlagNoPermBits. If this bit is set, the default permissions for that OS should be used.

However, what is happening in the code (at least to my limited knowledge) is:
- The FlagNoPermBits is set correctly on the server where the new file is created.
- On the receiving end, the ignorePermissions setting (belonging to the share) is used instead of checking whether the FlagNoPermBits are set on the files which are being received.
- This will give the same results as the intended behaviour except when you have a mixture of ignorePermissions settings on the servers. In that case, you will get very unpredictable permission settings.

### CODE REVIEW

**Comment by calmh** [comment]
@st-jenkins test this please

**Comment by calmh** [comment]
I don't think #400 is the right reference here?

**Comment by calmh** [comment]
Ah, I see you mean the issue in syncthing-android. Either create a "duplicate" on our side, or just let this be a reference to #1399 as well is fine I think.

I'll do some thinking, but I don't think this is the the entire solution. You're replacing checks for the `ignorePerms` setting with checking the bit, but you also have cases where updates are coming in from boxes with the full set of permissions, to a device with `ignorePerms` set.

The flag comparison is usually expressed as `if someVariable&protocol.FlagWhatever != 0` in the rest of the code.

**Comment by mogwa1** [comment]
I opened a new issue on the correct repository this time (sorry, it was a bit late last night...).

Ok, I see what you mean: it's probably a good idea to also ignore the permissions on the receiving side.  However, in that case, you might want to mention that behaviour in the documentation (or the original announcement) because there is currently no mention of it.  Currently, the intended behaviour is quite confusing.

Thanks for the tip on flag comparison.  I was looking for an example in the code, but couldn't find one straight away (and I'm not a bit tweaking guy anyway).

I guess the solution would be to check for either flag: ignorePerms OR FlagNoPermBits.  If you want, I can implement it in that way, and do some more checks?

File: internal/model/rwfolder.go [hunk 1]
```diff
@@ -109,6 +109,13 @@ func newRWFolder(m *Model, shortID uint64, cfg config.FolderConfiguration) *rwFo
 	}
 }
 
+// Helper function to check whether either the ignorePerm flag has been 
+// set on the local host or the FlagNoPermBits has been set on the file/dir
+// which is being pulled.
+func (p *rwFolder) ignorePermissions(file protocol.FileInfo) bool {
```
Comment (AudriusButkevicius) [inline internal/model/rwfolder.go]: You should perhaps do this on a file, so that you could do if file.IgnorePermissions()?

Comment (calmh) [inline internal/model/rwfolder.go]: I think this is fine since it needs to check `p.ignorePerms` as well which is not accessible from `FileInfo`.

**Comment by calmh** [comment]
@st-jenkins run integration tests please

**Comment by calmh** [comment]
Merged by cherry-pick-resolve-and-clean-up :)

**Comment by mogwa1** [comment]
Aha, I just logged in to make a similar comment about `p.ignorePerms`, but I see that the commit has meanwhile been accepted.  :-)  I just wanted to add that I ran several tests yesterday evening, using a setup of 3 hosts: 1 windows machine and 2 linux machines.  The proposed changes had the expected effects for all possible combinations of `ignorePerms` and several different umasks (on the linux hosts).
It looks like I can now completely switch over to syncthing.

Anyway, I'm glad to have been able to contribute!  Thanks for the quick responses!

**Pull request closed without merging**